### PR TITLE
Vectorize audio resampling for ARM NEON.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

## Category:

 **Other** Performance optimization (main purpose)
**Bug fix** (additional, in the same file)

## Description:
Implements vectorized single-channel audio resampling for ARM NEON.

## Additional information:
Vectorization is done differently than on SSE (pairwise loads are faster on ARM).
Other changes include improved handling of floor function.
Bug fix: SSE implementation used rounding instead of truncation - fixed.

### Affected modules and functionalities:
Audio resampling (audio decoder).

### Tests
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2651

### CODE REVIEW

File: dali/kernels/signal/resampling.h [hunk 1]
```diff
@@ -37,25 +43,63 @@ inline double Hann(double x) {
...
+    int32x4_t i = vcvtq_s32_f32(fi);
+    float32x4_t fifloor = vcvtq_f32_s32(i);
+    float32x4_t di = vsubq_f32(fi, fifloor);
+    int idx[4] = {
+        vgetq_lane_s32(i, 0),
+        vgetq_lane_s32(i, 1),
+        vgetq_lane_s32(i, 2),
+        vgetq_lane_s32(i, 3)
+    };
+    float32x2_t c0 = vld1_f32(&lookup[idx[0]]);
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:88]: :+1:

File: dali/kernels/signal/resampling.h [hunk 2]
```diff
@@ -132,11 +181,27 @@ struct Resampler {
...
+        float32x4_t f4 = vdupq_n_f32(0);
+
+        float32x4_t x4 = vaddq_f32(vdupq_n_f32(i - in_pos), _0123);
+
+        for (; i + 3 <= i1; i += 4) {
+            float32x4_t w4 = window(x4);
+            f4 = vfmaq_f32(f4, vld1q_f32(in_block_ptr + i), w4);
+            x4 = vaddq_f32(x4, vdupq_n_f32(4));
+        }
+        // Reduce elements in f4
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:199]: You can add the same comment in L214 - it was unclear for me why is happening there.

File: dali/kernels/signal/resampling.h [hunk 3]
```diff
@@ -132,11 +181,27 @@ struct Resampler {
...
         if (i0 + in_block_i < 0)
           i0 = -in_block_i;
-        if (i1 + in_block_i >= n_in)
-          i1 = n_in - 1 - in_block_i;
+        if (i1 + in_block_i > n_in)
+          i1 = n_in - in_block_i;
         float f = 0;
         int i = i0;
 
+#ifdef __ARM_NEON
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:189]: Maybe we can extract this vectorized parts into separate functions, instead having a lot of variants for different architectures in one body.

File: dali/kernels/signal/resampling.h [hunk 4]
```diff
@@ -105,6 +147,49 @@ struct Resampler {
...
 
+#if defined(__ARM_NEON)
+  inline float filter_vec(int &i_ref, float in_pos, int i1, const float *in) const {
+    const float32x4_t _0123 = vsetq_f32(0, 1, 2, 3);
+    float32x4_t f4 = vdupq_n_f32(0);
+
+    int i = i_ref;
+    float32x4_t x4 = vaddq_f32(vdupq_n_f32(i - in_pos), _0123);
+
+    for (; i + 3 <= i1; i += 4) {
```
Comment (JanuszL) [inline dali/kernels/signal/resampling.h:158]: Shouldn't you change that as well like in L174?

Comment (mzient) [reply dali/kernels/signal/resampling.h:158]: True.

**Review by JanuszL — APPROVED** [review]

File: dali/kernels/signal/resampling.h [hunk 5]
```diff
@@ -37,25 +40,63 @@ inline double Hann(double x) {
   return 0.5 * (1 + std::cos(x * M_PI));
 }
 
+#ifdef __ARM_NEON
+
+inline float32x4_t vsetq_f32(float x0, float x1, float x2, float x3) {
+    float32x4_t x;
```
Comment (prak-nv) [inline dali/kernels/signal/resampling.h:46]: Very little nitpick, could be `float32x4_t x = vdubpq_n_f32(x0);`

File: dali/kernels/signal/resampling.h [hunk 6]
```diff
@@ -37,25 +40,63 @@ inline double Hann(double x) {
...
+    float32x4_t next = w.val[1];
+    return vfmaq_f32(curr, di, vsubq_f32(next, curr));
+  }
+#endif
+
 #ifdef __SSE2__
   inline __m128 operator()(__m128 x) const {
     __m128 fi = _mm_add_ps(x * _mm_set1_ps(scale), _mm_set1_ps(center));
-    __m128i i = _mm_cvtps_epi32(fi);
+    __m128i i = _mm_cvttps_epi32(fi);
```
Comment (prak-nv) [inline dali/kernels/signal/resampling.h:99]: Was doing conversion without truncate an issue before?

Comment (mzient) [reply dali/kernels/signal/resampling.h:99]: Yes, it was a bug - I detected it now that there's a slight difference between ARM and SSE and dug deeper to find out that it's SSE that's wrong.

**Pull request merged**